collector/workitem.go

```diff
@@ -0,0 +1,337 @@
+		errors = append(errors, err)
+	}
+	users = append(users, collectSpaceCollaboratorUsers(sc)...)
+
+	resolved, err := resolveAllUsers(ctx, c, SliceUniq(users), sc.Data)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	if len(errors) > 0 {
+		return resolved, values, multiError{Message: "Errors during notification resolving", Errors: errors}
+	}
+
+	return resolved, values, nil
+}
+
+func resolveAllUsers(ctx context.Context, c *api.Client, users []uuid.UUID, collaborators []*api.UserData) ([]Receiver, error) {
+	var resolved []Receiver
+
+	for _, u := range users {
+		found := false
+		for _, c := range collaborators {
+			if u.String() == *c.ID {
+				found = true
+				if c.Attributes.Email != nil {
+					user := Receiver{EMail: *c.Attributes.Email}
+					if c.Attributes.FullName != nil {
+						user.FullName = *c.Attributes.FullName
+					}
+					resolved = append(resolved, user)
+				}
+			}
+		}
+		if !found {
+			c, err := wit.GetUser(ctx, c, u)
+			if err == nil {
+				if c.Data.Attributes.Email != nil {
+					user := Receiver{EMail: *c.Data.Attributes.Email}
+					if c.Data.Attributes.FullName != nil {
+						user.FullName = *c.Data.Attributes.FullName
+					}
+					resolved = append(resolved, user)
+				}
+			}
+		}
+	}
+
+	return resolved, nil
+}
+
+func collectSpaceCollaboratorUsers(cl *api.UserList) []uuid.UUID {
+	var users []uuid.UUID
+	for _, c := range cl.Data {
+		cID, err := uuid.FromString(*c.ID)
+		if err == nil {
+			users = append(users, cID)
+		}
+	}
+	return users
+}
+
+func collectCommentsUsers(cl *api.CommentList) []uuid.UUID {
+	var users []uuid.UUID
+
+	for _, c := range cl.Data {
+		if c.Relationships.CreatedBy != nil {
+			users = append(users, *c.Relationships.CreatedBy.Data.ID)
+		}
+	}
+
+	return users
+}
+
+func collectCommentUsers(c *api.CommentSingle) []uuid.UUID {
+	var users []uuid.UUID
+
+	if c.Data.Relationships.CreatedBy != nil {
+		users = append(users, *c.Data.Relationships.CreatedBy.Data.ID)
+	}
+
+	return users
+}
+
+func collectWorkItemUsers(wi *api.WorkItemSingle) []uuid.UUID {
+	var users []uuid.UUID
+
+	if wi.Data.Relationships.Creator != nil && wi.Data.Relationships.Creator.Data != nil {
+		creatorID, err := uuid.FromString(*wi.Data.Relationships.Creator.Data.ID)
+		if err == nil {
+			users = append(users, creatorID)
+		}
+	}
+
+	if wi.Data.Relationships.Assignees != nil && wi.Data.Relationships.Assignees.Data != nil {
+		for _, assignee := range wi.Data.Relationships.Assignees.Data {
+			assigneeID, err := uuid.FromString(*assignee.ID)
+			if err == nil {
+				users = append(users, assigneeID)
+			}
+		}
+	}
+	return users
+}
+
+func collectSpaceUsers(space *api.SpaceSingle) []uuid.UUID {
+	var users []uuid.UUID
+
+	if space.Data.Relationships.OwnedBy != nil {
+		users = append(users, *space.Data.Relationships.OwnedBy.Data.ID)
+	}
+	return users
+}
+
+// SliceUniq removes duplicate values in given slice
+func SliceUniq(s []uuid.UUID) []uuid.UUID {
+	for i := 0; i < len(s); i++ {
```

Comment:
I needed to go up to 10.000 items for it to even register a difference between the two, but yes. With 100.000 map is clearly faster. I was thinking map would kill perf on small, but it's only stack allocated. :)